lib/blocks.js

```diff
@@ -841,75 +859,170 @@ var blockStarts = [
-            parser.addChild("table_cell", parser.offset);
-
-            parser.tip._string_content = cells[i].trim();
-
-            parser.advanceOffset(cells[i].length + 1);
+            parser.tip.appendChild(cell);
         }
 
+        // Mark the rest of the line as read
+        parser.advanceOffset(
+            parser.currentLine.length - parser.offset,
+            false
+        );
+
         return 2;
     }
 ];
 
-const parseDelimiterRow = function(row) {
-    if (row.indexOf("|") === -1) {
+const parseTableRow = function(line, startAt) {
+    // This is attempting to replicate row_from_string from GitHub's Commonmark fork. That function can be found here:
+    // https://github.com/github/cmark-gfm/blob/587a12bb54d95ac37241377e6ddc93ea0e45439b/extensions/table.c#L189
+
+    let cells = [];
+
+    let expectMoreCells = true;
+
+    // Start at the current parser position
+    let offset = startAt;
+
+    // Read past the optional leading pipe
+    offset += scanTableCellEnd(line, offset);
+
+    while (offset < line.length && expectMoreCells) {
+        const cellLength = scanTableCell(line, offset);
+        const pipeLength = scanTableCellEnd(line, offset + cellLength);
+
+        if (cellLength > 0 || pipeLength > 0) {
+            // We're guaranteed to have found a cell because we either found some cell content (cellLength > 0) or
+            // we found an empty cell with a pipe (cellLength == 0 && pipeLength > 0)
+            const cellContents = unescapePipes(line.substring(offset, offset + cellLength));
+
+            cells.push({
+                contents: cellContents,
+                start: offset,
+                end: offset + cellLength,
+            });
+
+            offset += cellLength + pipeLength;
+        }
+
+        if (pipeLength > 0) {
+            expectMoreCells = true;
+        } else {
+            // We've read the last cell, so check if we've reached the end of the row
+            const rowEndLength = scanTableRowEnd(line, offset);
+
+            // TODO there's something in the GH code about the row being part of a preceding paragraph which we may not
+            // need. I think it has to do with GH's behaviour where there's no double-newline needed between a
+            // paragraph and a table which we don't support in Marked or Commonmark.
+
+            if (rowEndLength === -1) {
+                // There's more text after this, so this isn't a table row
+            } else {
+                offset += rowEndLength;
+            }
+
+            expectMoreCells = false;
+        }
+    }
+
+    if (offset === line.length) {
+        // We've read the whole line, so it's a valid row
+        return cells;
+    } else {
+        // There's unhandled text here, so it's not actually a table row
         return null;
     }
+};
 
-    const cells = parseTableCells(row).map((cell) => cell.trim());
+const reTableCell = new RegExp("^([\\\\]" + ESCAPABLE + "|[^|\r\n])+");
```

Comment:
This reads any number of characters until it hits a newline or a pipe that isn't escaped (`\|`). It still uses a regular expression because the C way of doing this involves [a lookup table generated with a lexer generator](https://github.com/github/cmark-gfm/blob/master/extensions/ext_scanners.c#L229-L245)